src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -2998,11 +2983,25 @@ internal void WriteByte(byte b)
             _outBuff[_outBytesUsed++] = b;
         }
 
+        internal Task WriteByteSpan(ReadOnlySpan<byte> span, bool canAccumulate = true, TaskCompletionSource<object> completion = null)
+        {
+            return WriteByteArray(span, span.Length, 0, canAccumulate, completion);
+        }
+
+        internal Task WriteByteArray(byte[] b, int len, int offsetBuffer, bool canAccumulate = true, TaskCompletionSource<object> completion = null)
+        {
+            return WriteByteArray(ReadOnlySpan<byte>.Empty, len, offsetBuffer, canAccumulate, completion, b);
+        }
+
         //
         // Takes a byte array and writes it to the buffer.
         //
-        internal Task WriteByteArray(byte[] b, int len, int offsetBuffer, bool canAccumulate = true, TaskCompletionSource<object> completion = null)
+        internal Task WriteByteArray(ReadOnlySpan<byte> b, int len, int offsetBuffer, bool canAccumulate = true, TaskCompletionSource<object> completion = null, byte[] array = null)
```

Comment:
The flow of this function is dependent on `array` . It is a bit confusing, what is the source of data for this method.
If the byte[] array is defined then it is used to create a Span, else `b` is used. 
Will it make sense to create an overload of this function, one with ReadOnlySpan b and one which takes byte[] array ? From the looks of it, if `b` is passed in and `array` is passed in, then the data in array takes precedence, which means that the data in `b` will be ignored.

Reply:
If you pass in a span and no byte array then it will attempt to use only the span. 
If you pass in an array then the array will be used as the source of the span. 
if the span cannot be written into the current packet then the remaining contents of the span are "promoted" to a heap allocated array so they can be part of the async closure that will used to callback into the method to continue. 
Once you're got a full array the method will use that in preference to anything else so you can't be in a position where you accidentally pass a span and an array, array wins.

Reply:
@Wraith2  This is what I want to prevent in the method. Having to read through code to understand what is the data source :smile:

Reply:
If that's clear enough I can put it into the code at the top of the function, would that work?

Reply:
Yeah, make a comment on top of the function explaining this. Thanks.

Reply:
I've added a coment to the header of the function explaining how it's used. I've also changed it from internal to private because i'd already provided wrappers for both array and span only, this way you have to be in the same file as the confusing parametered version to call it and at that point you chould read the comment. I also decided to rename it to WriteBytes because it's more accurate. Externally you have to call WriteByteArray or WriteByteSpan.